instrumentation/jsf/jsf-testing-common/src/main/groovy/BaseJsfTest.groovy

```diff
@@ -0,0 +1,227 @@
+import org.jsoup.Jsoup
+
+abstract class BaseJsfTest extends AbstractHttpServerTest<Server> {
+
+  @Override
+  Server startServer(int port) {
+    List<String> configurationClasses = new ArrayList<>()
+    Collections.addAll(configurationClasses, WebAppContext.getDefaultConfigurationClasses())
+    configurationClasses.add(AnnotationConfiguration.getName())
+
+    WebAppContext webAppContext = new WebAppContext()
+    webAppContext.setContextPath(getContextPath())
+    webAppContext.setConfigurationClasses(configurationClasses)
+    // set up test application
+    webAppContext.setBaseResource(Resource.newSystemResource("test-app"))
+    // add additional resources for test app
+    Resource extraResource = Resource.newSystemResource("test-app-extra")
+    if (extraResource != null) {
+      webAppContext.getMetaData().addWebInfJar(extraResource)
+    }
+    webAppContext.getMetaData().getWebInfClassesDirs().add(Resource.newClassPathResource("/"))
+
+    def jettyServer = new Server(port)
+    jettyServer.connectors.each {
+      it.setHost('localhost')
+    }
+
+    jettyServer.setHandler(webAppContext)
+    jettyServer.start()
+
+    return jettyServer
+  }
+
+  @Override
+  void stopServer(Server server) {
+    server.stop()
+    server.destroy()
+  }
+
+  @Override
+  String getContextPath() {
+    return "/jetty-context"
+  }
+
+  void addServlet(ServletContextHandler servletContext, String path, Class<Servlet> servlet) {
+    servletContext.addServlet(servlet, path)
+  }
+
+  def "test hello"() {
+    setup:
+    def url = HttpUrl.get(address.resolve("hello.jsf")).newBuilder().build()
+    def request = request(url, "GET", null).build()
+    Response response = client.newCall(request).execute()
+
+    expect:
+    response.code() == 200
+    response.body().string().trim() == "Hello"
+
+    and:
+    assertTraces(1) {
+      trace(0, 1) {
+        basicSpan(it, 0, getContextPath() + "/hello.xhtml", null)
+      }
+    }
+  }
+
+  def "test greeting"() {
+    // we need to display the page first before posting data to it
+    setup:
+    def url = HttpUrl.get(address.resolve("greeting.jsf")).newBuilder().build()
+    def request = request(url, "GET", null).build()
+    Response response = client.newCall(request).execute()
+    def doc = Jsoup.parse(response.body().string())
+
+    expect:
+    response.code() == 200
+    doc.selectFirst("title").text() == "Hello, World!"
+
+    and:
+    assertTraces(1) {
+      trace(0, 1) {
+        basicSpan(it, 0, getContextPath() + "/greeting.xhtml", null)
+      }
+    }
+
+    when:
+    // extract parameters needed to post back form
+    def viewState = doc.selectFirst("[name=javax.faces.ViewState]")?.val()
+    def formAction = doc.selectFirst("#app-form").attr("action")
+    def jsessionid = formAction.substring(formAction.indexOf("jsessionid=") + "jsessionid=".length())
+
+    then:
+    viewState != null
+    jsessionid != null
+
+    when:
+    TEST_WRITER.clear()
```

Comment:
Nit: I think that you can move this line just after the `assertTraces` call, this way the whole trace logic for this test step would be wrapped in one block.

Reply:
moved